testsuite/features/support/api_test.rb

```diff
@@ -0,0 +1,585 @@
+  end
+
+  ## "system.custominfo" namespace
+
+  def system_custominfo_create_key(value, desc)
+    @connection.call('system.custominfo.createKey', sessionKey: @sid, keyLabel: value, keyDescription: desc)
+  end
+
+  ## "system.provisioning.powermanagement" namespace
+
+  def system_provisioning_powermanagement_list_types
+    @connection.call('system.provisioning.powermanagement.listTypes', sessionKey: @sid)
+  end
+
+  def system_provisioning_powermanagement_get_details(server)
+    @connection.call('system.provisioning.powermanagement.getDetails', sessionKey: @sid, serverId: server)
+  end
+
+  def system_provisioning_powermanagement_get_status(server)
+    @connection.call('system.provisioning.powermanagement.getStatus', sessionKey: @sid, serverId: server)
+  end
+
+  def system_provisioning_powermanagement_set_details(server, data)
+    @connection.call('system.provisioning.powermanagement.setDetails', sessionKey: @sid, serverId: server, data: data)
+  end
+
+  def system_provisioning_powermanagement_power_on(server)
+    @connection.call('system.provisioning.powermanagement.powerOn', sessionKey: @sid, serverId: server)
+  end
+
+  def system_provisioning_powermanagement_power_off(server)
+    @connection.call('system.provisioning.powermanagement.powerOff', sessionKey: @sid, serverId: server)
+  end
+
+  def system_provisioning_powermanagement_reboot(server)
+    @connection.call('system.provisioning.powermanagement.reboot', sessionKey: @sid, serverId: server)
+  end
+
+  ## "system.scap" namespace
+
+  def system_scap_list_xccdf_scans(server)
+    @connection.call('system.scap.listXccdfScans', @sid, serverId: server)
+  end
+
+  ## "user" namespace
+
+  def user_list_users
+    @connection.call('user.listUsers', sessionKey: @sid)
+  end
+
+  def user_list_roles(user)
+    @connection.call('user.listRoles', sessionKey: @sid, login: user)
+  end
+
+  def user_create(user, password, first, last, email)
+    @connection.call('user.create', sessionKey: @sid, desiredLogin: user, desiredPassword: password, firstName: first, lastName: last, email: email)
+  end
+
+  def user_delete(user)
+    @connection.call('user.delete', sessionKey: @sid, login: user)
+  end
+
+  def user_add_role(user, role)
+    @connection.call('user.addRole', sessionKey: @sid, login: user, role: role)
+  end
+
+  def user_remove_role(user, role)
+    @connection.call('user.removeRole', sessionKey: @sid, login: user, role: role)
+  end
+
+  def user_get_details(user)
+    @connection.call('user.getDetails', sessionKey: @sid, login: user)
+  end
+end
+
+# Derived class for XML-RPC test
+class APITestXMLRPC < APITestBase
+  def initialize(host)
+    @connection = XmlrpcClient.new(host)
+  end
+
+  def is_date?(attribute)
+    attribute.class == XMLRPC::DateTime
+  end
+
+  def date_now
+    now = Time.now
+    XMLRPC::DateTime.new(now.year, now.month, now.day, now.hour, now.min, now.sec)
+  end
```

Comment:
I'm not sure about adding that logic at that layer.
What's the special need of this class that should handle API calls to provide a Date Helper? :thinking:

Reply:
it's because the dates are handled differently in XML-RPC and HTTP. It's due to the fact that json file format cannot handle dates, while XML-RPC can because it's based on XML. This is why HTTP sees all dates as simple strings (with all formatting issues that might arise). If you ask me, that's a regression.

Reply:
I would have a class or tool, to provide date_now depending your needs (xmlrpc, http).
Otherwise we are combining responsibilities in a unique classes.
"[single-responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle)"

But I will not made a strong opinion on that. As you feel, just sharing thoughs.

Reply:
Well, implementing practical details in derived classes is a basic principle of object-oriented programming... I can even point you to pages in reference books about that.